go/plugins/ollama/ollama.go

```diff
@@ -0,0 +1,228 @@
+func (g *generator) generate(ctx context.Context, input *ai.GenerateRequest, cb func(context.Context, *ai.GenerateResponseChunk) error) (*ai.GenerateResponse, error) {
+	// Step 1: Combine parts from all messages into a single payload slice
+	var messages []map[string]string
+
+	// Add all messages to history field.
+	for _, m := range input.Messages {
+		message, err := convertParts(m.Role, m.Content)
+		if err != nil {
+			return nil, fmt.Errorf("error converting message parts: %v", err)
+		}
+		messages = append(messages, message)
+	}
+	fmt.Println("should stream", cb != nil)
+	stream := cb != nil
+	payload := ollamaRequest{
+		Messages: messages,
+		Model:    g.model,
+		Stream:   stream,
+	}
+	client := &http.Client{
+		Timeout: time.Second * 30,
+	}
+	payloadBytes, err := json.Marshal(payload)
+	req, err := http.NewRequest("POST", g.serverAddress+"/api/chat", bytes.NewBuffer(payloadBytes))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send request: %v", err)
+	}
+	defer resp.Body.Close()
+	if cb == nil {
+		// Existing behavior for non-streaming responses
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read response body: %v", err)
+		}
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("server returned non-200 status: %d, body: %s", resp.StatusCode, body)
+		}
+		response, err := translateResponse(body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse response: %v", err)
+		}
+		return response, nil
+	} else {
+		// Handle streaming response here
+		scanner := bufio.NewScanner(resp.Body) // Create a scanner to read lines
+		for scanner.Scan() {
+			line := scanner.Text()
+
+			chunk, err := translateChunk(line)
+			if err != nil {
+				// Handle parsing error (log, maybe send an error candidate?)
+				return nil, fmt.Errorf("error translating chunk: %v", err)
+			}
+			cb(ctx, chunk)
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, fmt.Errorf("error reading stream: %v", err)
+		}
+		// Handle end of stream (optional: send a final candidate to signal completion)
+	}
+	//Return an empty generate response, since we use callback for streaming
+	return &ai.GenerateResponse{}, nil
+}
+
+// convertParts serializes a slice of *ai.Part into a a map (represent Ollama message type)
+func convertParts(role ai.Role, parts []*ai.Part) (map[string]string, error) {
+	roleMapping := map[ai.Role]string{
+		ai.RoleUser:   "user",
+		ai.RoleModel:  "assistant",
+		ai.RoleSystem: "system",
+		// Add more mappings as needed
+	}
+
+	partMap := map[string]string{}
+	for _, part := range parts {
+		partMap["role"] = roleMapping[role]
+		switch {
+		case part.IsText():
+			partMap["content"] = part.Text
```

Comment:
What happens if there are multiple parts? Shouldn't we concatenate these, or make a list, or something?

Reply:
Ah thats true, I only handled the text case here where there is only 1 part but that's a good point, I updated the logic to make it generic.